### ISSUE TITLE
Truncating Batches is Dangerous

### ISSUE THREAD

**Issue by profPlum** [Lightning-AI/lightning#18633]
### Bug description

It seems that pytorch-lightning works by truncating the last batch in the epoch if there aren't enough remaining training examples to fill it. This seems fine for small batches like 32-64 or something. 

But for very big batches (**which are encouraged by the batch size finder feature**), it becomes very inappropriate b/c you might have a batch size of like 100k which naturally warrants a correspondingly very high learning rate. 

**Except in practice if there are say 200,040 samples in the training dataset, then you have batch sizes schedule like this repeated: 100,000, 100,000, 40.** The last batch is almost trivially sized but could still be given a very high learning rate since LR is set for batches that are supposed to be 100k big. This causes very unstable training and/or confusing LR finder results (e.g. LR~=1e-8 recommended for 100k batchsize & LR~=1e-4 recommended for 100,020).

My suggestion: Rather than truncating the batch (which can be quite dangerous/unstable depending on batch size!), instead make the dataset 'cyclic' so if there are not enough samples for the last batch take them from the next epoch.

After all, [epochs have become passé](https://arxiv.org/abs/1906.06669) as a measure of training progress... Recently people have begun to think of gradient steps (_still with cyclic data_) rather than epochs as primary measure of progress.

### What version are you seeing the problem on?

v1.9

**Comment by profPlum** [comment]
@awaelchli I stand corrected, I thought that was the reason. But apparently LearningRateFinder is just very erratic... It's recommendations will sway between 1e-4 & 1e-8, based on model initialization.

And of course you're right that the thing I mentioned about truncation is related to pytorch itself... Sorry for confusion.

**Comment by awaelchli** [comment]
@profPlum If you believe there is a systematic problem with the LR-finder, feel free to open a separate issue for it. Nobody has worked on any significant improvements for the LRFinder algorithm in over 2 years so I wouldn't expect it to work exceptionally well in all cases. 

It seems that the confusion bout dropped samples is resolved, so I'm closing this now.
